hi-ml/src/health_ml/utils/reports.py

```diff
@@ -0,0 +1,165 @@
+        self.report_title = title
+        self.output_folder = output_folder
+
+        report_folder = Path(output_folder)  # / title.lower().replace(" ", "_")
+        report_folder.mkdir(exist_ok=True, parents=True)
+
+        self.report_folder = report_folder
+        self.report_path_html = report_folder / (title.lower().replace(" ", "_") + '.html')
+        self.report_path_pdf = report_folder / (title.lower().replace(" ", "_") + '.pdf')
+        self.report_html = ""
+        self.template = ""
+        self.template_path = self._create_template()
+
+        self.env = jinja2.Environment(
+            loader=jinja2.FileSystemLoader('/')
+        )
+        self.render_kwargs: Dict[str, Any] = {"title": title}
+
+    @staticmethod
+    def _remove_html_end(report_stream):
+        return report_stream.replace("</p>\n</div>\n</body>\n</html>", "")
+
+    def _create_template(self):
+        template_path = self.report_folder / "template.html"
+        template_path.touch(exist_ok=True)
+
+        self.template += """<!DOCTYPE html>
+<html lang="en">
+<head>
+<meta charset="UTF-8">
+<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css" rel="stylesheet">
+<style type="text/css">
+    @page body{
+        size: A4 portrait;
+        @frame header_frame {
+            -pdf-frame-content: header_content;
+            left: 50pt; width: 512pt; top: 50pt; height: 40pt;
+        }
+        @frame content_frame {
+            left: 50pt; width: 512pt; top: 90pt; height: 632pt;
+        }
+        @frame footer_frame {
+            -pdf-frame-content: footer_content;
+            left: 50pt; width: 512pt; top: 772pt; height: 20pt;
+        }
+    }
+    tbody td {
+        text-align: center;
+        vertical-align: middle}
+    thead th {
+        text-align: center;}
+</style>
+<title> {{title}} </title>
+</head>
+<body>
+<div class="container-fluid justify-content-center align-items-center">
+<div class="container">
+<h1> {{title}} </h1>
+</div>
+<p>
+</p>
+</div>
+</body>
+</html>"""
+
+        return template_path
+
+    def add_table(self, df: pd.DataFrame):
+
+        self.template = self._remove_html_end(self.template)
+
+        num_existing_tables = self.template.count("table.to_html")
+        table_key = f"{TABLE_KEY_HTML}_{num_existing_tables}"  # starts at zero
+
+        self.template += """<div class="container" >
+{% for table in """ + table_key + """ %}
+    {{ table.to_html(classes=[ "table"], justify="center") | safe }}
+{% endfor %}
+</div>
+<br>
+</p>
+</div>
+</body>
+</html>"""
+
+        self.render_kwargs.update({table_key: [df]})
+
+    def add_image(self, image_path: str):
+        if image_path.startswith(str(self.report_folder)):
+            img_path_html = Path(image_path).relative_to(self.report_folder)
+        else:
+            img_path_html = Path(image_path)
+
+        image_path_pdf = Path(image_path)
+
+        self.template = self._remove_html_end(self.template)
```

Comment:
There's some DRY opportunity here. the code repeatedly does "move html_end, add to template, add html_end"